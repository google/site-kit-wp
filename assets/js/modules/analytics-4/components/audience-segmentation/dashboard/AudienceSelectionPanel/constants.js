/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
