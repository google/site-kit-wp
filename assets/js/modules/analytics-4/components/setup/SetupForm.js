/**
 * Analytics Setup form.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useCallback, useEffect } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { SpinnerButton } from 'googlesitekit-components';
import {
	FORM_SETUP,
	EDIT_SCOPE,
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
import { isPermissionScopeError } from '../../../../util/errors';
import SetupFormFields from './SetupFormFields';
import StoreErrorNotices from '../../../../components/StoreErrorNotices';

import useViewContext from '../../../../hooks/useViewContext';
import { trackEvent } from '../../../../util';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import SetupEnhancedConversionTrackingNotice from '../../../../components/conversion-tracking/SetupEnhancedConversionTrackingNotice';

export default function SetupForm( { finishSetup } ) {
	const hasEditScope = useSelect( ( select ) =>
		select( CORE_USER ).hasScope( EDIT_SCOPE )
	);
	const autoSubmit = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue( FORM_SETUP, 'autoSubmit' )
	);
	const canSubmitChanges = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).canSubmitChanges()
	);
	const isSaving = useSelect(
		( select ) =>
			select( MODULES_ANALYTICS_4 ).isDoingSubmitChanges() ||
			select( CORE_LOCATION ).isNavigating()
	);
	const viewContext = useViewContext();

	const { setValues } = useDispatch( CORE_FORMS );
	const { submitChanges } = useDispatch( MODULES_ANALYTICS_4 );
	const { setConversionTrackingEnabled, saveConversionTrackingSettings } =
		useDispatch( CORE_SITE );

	const isEnhancedMeasurementEnabled = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue(
			ENHANCED_MEASUREMENT_FORM,
			ENHANCED_MEASUREMENT_ENABLED
		)
	);

	const submitForm = useCallback(
		async ( event ) => {
			event.preventDefault();
			// Disable autoSubmit unconditionally to prevent
			// automatic invocation more than once.
			setValues( FORM_SETUP, { autoSubmit: false } );

			const { error } = await submitChanges();

			if ( isPermissionScopeError( error ) ) {
				setValues( FORM_SETUP, { autoSubmit: true } );
			}

			if ( ! error ) {
				setConversionTrackingEnabled( true );
				await saveConversionTrackingSettings();

				if ( isEnhancedMeasurementEnabled === true ) {
					await trackEvent(
						`${ viewContext }_analytics`,
						'ga4_setup_enhanced_measurement_enabled'
					);
				}
				finishSetup();
			}
		},
		[
			finishSetup,
			isEnhancedMeasurementEnabled,
			setConversionTrackingEnabled,
			saveConversionTrackingSettings,
			setValues,
			submitChanges,
			viewContext,
		]
	);

	// If the user lands back on this component with autoSubmit and the edit scope,
	// resubmit the form.
	useEffect( () => {
		if ( autoSubmit && hasEditScope ) {
			submitForm( { preventDefault: () => {} } );
		}
	}, [ hasEditScope, autoSubmit, submitForm ] );

	return (
		<form
			className="googlesitekit-analytics-setup__form"
			onSubmit={ submitForm }
		>
			<StoreErrorNotices
				moduleSlug="analytics-4"
				storeName={ MODULES_ANALYTICS_4 }
			/>
			<SetupFormFields />

			<SetupEnhancedConversionTrackingNotice
				message={ __(
					'To track how visitors interact with your site, Site Kit will enable enhanced conversion tracking. You can always disable it in settings.',
					'google-site-kit'
				) }
			/>

			<div className="googlesitekit-setup-module__action">
				<SpinnerButton
					disabled={ ! canSubmitChanges || isSaving }
					isSaving={ isSaving }
				>
					{ __( 'Complete setup', 'google-site-kit' ) }
				</SpinnerButton>
			</div>
		</form>
	);
}

SetupForm.propTypes = {
	finishSetup: PropTypes.func,
};

SetupForm.defaultProps = {
	finishSetup: () => {},
};
