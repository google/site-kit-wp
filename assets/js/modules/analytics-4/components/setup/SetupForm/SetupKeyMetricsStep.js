/**
 * Analytics Setup form.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { SpinnerButton } from 'googlesitekit-components';
import {
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	MODULES_ANALYTICS_4,
} from '../../../datastore/constants';
import { CORE_FORMS } from '../../../../../googlesitekit/datastore/forms/constants';
import { CORE_LOCATION } from '../../../../../googlesitekit/datastore/location/constants';
import SetupFormFields from '../SetupFormFields';
import StoreErrorNotices from '../../../../../components/StoreErrorNotices';

import useViewContext from '../../../../../hooks/useViewContext';
import { trackEvent } from '../../../../../util';
import { CORE_SITE } from '../../../../../googlesitekit/datastore/site/constants';

export default function SetupKeyMetricsStep( { finishSetup } ) {
	const viewContext = useViewContext();

	const isSaving = useSelect(
		( select ) =>
			select( MODULES_ANALYTICS_4 ).isDoingSubmitChanges() ||
			select( CORE_LOCATION ).isNavigating()
	);

	const canSubmitChanges = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).canSubmitChanges()
	);
	const { submitChanges } = useDispatch( MODULES_ANALYTICS_4 );
	const { setConversionTrackingEnabled, saveConversionTrackingSettings } =
		useDispatch( CORE_SITE );

	const isEnhancedMeasurementEnabled = useSelect( ( select ) =>
		select( CORE_FORMS ).getValue(
			ENHANCED_MEASUREMENT_FORM,
			ENHANCED_MEASUREMENT_ENABLED
		)
	);

	const submitForm = useCallback(
		async ( event ) => {
			event.preventDefault();

			const { error } = await submitChanges();

			if ( ! error ) {
				setConversionTrackingEnabled( true );
				await saveConversionTrackingSettings();

				if ( isEnhancedMeasurementEnabled === true ) {
					await trackEvent(
						`${ viewContext }_analytics`,
						'ga4_setup_enhanced_measurement_enabled'
					);
				}
				finishSetup();
			}
		},
		[
			finishSetup,
			isEnhancedMeasurementEnabled,
			setConversionTrackingEnabled,
			saveConversionTrackingSettings,
			submitChanges,
			viewContext,
		]
	);

	return (
		<form
			className="googlesitekit-analytics-setup__form"
			onSubmit={ submitForm }
		>
			<StoreErrorNotices
				moduleSlug="analytics-4"
				storeName={ MODULES_ANALYTICS_4 }
			/>
			<SetupFormFields />

			<div className="googlesitekit-setup-module__action">
				<SpinnerButton
					disabled={ ! canSubmitChanges || isSaving }
					isSaving={ isSaving }
				>
					{ __( 'Complete setup', 'google-site-kit' ) }
				</SpinnerButton>
			</div>
		</form>
	);
}

SetupKeyMetricsStep.propTypes = {
	finishSetup: PropTypes.func,
};

SetupKeyMetricsStep.defaultProps = {
	finishSetup: () => {},
};
