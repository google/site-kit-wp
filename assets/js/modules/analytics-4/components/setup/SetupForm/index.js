/**
 * Analytics Setup form.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { HashRouter, Route, Switch, Redirect } from 'react-router-dom';

/**
 * Internal dependencies
 */
import ConnectAnalyticsStep from './ConnectAnalyticsStep';
import SetupKeyMetricsStep from './SetupKeyMetricsStep';

export default function SetupForm( { finishSetup } ) {
	return (
		<HashRouter>
			<Switch>
				<Route path="/connect-analytics">
					<ConnectAnalyticsStep />
				</Route>
			</Switch>
			<Route path="/setup-key-metrics">
				<SetupKeyMetricsStep finishSetup={ finishSetup } />
			</Route>
			<Redirect from="/" to="/connect-analytics" exact />
		</HashRouter>
	);
}

SetupForm.propTypes = {
	finishSetup: PropTypes.func,
};

SetupForm.defaultProps = {
	finishSetup: () => {},
};
