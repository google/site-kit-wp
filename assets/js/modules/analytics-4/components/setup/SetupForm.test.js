/**
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { act, fireEvent, render } from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	muteFetch,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	waitForDefaultTimeouts,
} from '../../../../../../tests/js/utils';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { MODULES_TAGMANAGER } from '../../../tagmanager/datastore/constants';
import {
	EDIT_SCOPE,
	ENHANCED_MEASUREMENT_ENABLED,
	ENHANCED_MEASUREMENT_FORM,
	FORM_SETUP,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY } from '../../../analytics-4/constants';
import * as fixtures from '../../datastore/__fixtures__';
import ga4ReportingTour from '../../../../feature-tours/ga4-reporting';
import SetupForm from './SetupForm';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';

const {
	accountSummaries,
	defaultEnhancedMeasurementSettings,
	webDataStreamsBatch,
	webDataStreams,
} = fixtures;
const accounts = accountSummaries.accountSummaries;
const properties = accounts[ 1 ].propertySummaries;
const accountID = accounts[ 1 ]._id;
const propertyID = properties[ 0 ]._id;
const webDataStreamID = webDataStreamsBatch[ propertyID ][ 0 ]._id;
const REGEX_REST_GA4_SETTINGS = new RegExp( '/analytics-4/data/settings' );
const REGEX_REST_DISMISS_ITEM = new RegExp(
	'^/google-site-kit/v1/core/user/data/dismiss-item'
);
const REGEX_REST_GA4_CREATE_PROPERTY = new RegExp(
	'/analytics-4/data/create-property'
);
const REGEX_REST_GA4_CREATE_WEBDATASTREAM = new RegExp(
	'/analytics-4/data/create-webdatastream'
);
const REGEX_REST_GA4_ACCOUNT_SUMMARIES = new RegExp(
	'/analytics-4/data/account-summaries'
);
const REGEX_REST_CONVERSION_TRACKING_SETTINGS = new RegExp(
	'^/google-site-kit/v1/core/site/data/conversion-tracking'
);

describe( 'SetupForm', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideSiteInfo( registry );
		provideUserAuthentication( registry );
		provideModules( registry, [ { slug: 'analytics-4', active: true } ] );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( CORE_SITE ).receiveGetConversionTrackingSettings( {
			enabled: false,
		} );
		muteFetch( REGEX_REST_CONVERSION_TRACKING_SETTINGS );
	} );

	it( 'renders the form correctly', async () => {
		registry.dispatch( CORE_SITE ).receiveGetConversionTrackingSettings( {
			enabled: true, // Hide notice for this case.
		} );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				{
					...defaultEnhancedMeasurementSettings,
					streamEnabled: false,
				},
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		const { container, getByText, waitForRegistry } = render(
			<SetupForm />,
			{
				registry,
			}
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Account' ) ).toBeInTheDocument();
		expect( getByText( 'Property' ) ).toBeInTheDocument();
		expect( getByText( 'Web Data Stream' ) ).toBeInTheDocument();
	} );

	it( 'submits the form upon pressing the CTA', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
			adsConversionID: '',
		} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		act( () => {
			// It doesn't seem possible to show the dropdown menu within the test by clicking on the dropdown in the usual way.
			// Therefore, we simulate the click on the dropdown menu item directly, despite it being hidden.
			fireEvent.click(
				getByRole( 'menuitem', {
					name: /Example Com/i,
					hidden: true,
				} )
			);
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		muteFetch( REGEX_REST_GA4_SETTINGS );

		fetchMock.post( REGEX_REST_DISMISS_ITEM, {
			status: 200,
			body: JSON.stringify( [
				ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
			] ),
		} );

		act( () => {
			fireEvent.click(
				getByRole( 'button', { name: /Complete setup/i } )
			);
		} );

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).toHaveFetchedTimes( 1, REGEX_REST_GA4_SETTINGS );

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries( {
			accountSummaries: accountSummaries.accountSummaries.map(
				( account ) => ( {
					...account,
					propertySummaries: [],
				} )
			),
			nextPageToken: null,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				fixtures.defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);
		registry.dispatch( CORE_FORMS ).setValues( ENHANCED_MEASUREMENT_FORM, {
			[ ENHANCED_MEASUREMENT_ENABLED ]: false,
		} );

		fetchMock.post(
			new RegExp(
				'^/google-site-kit/v1/modules/analytics-4/data/enhanced-measurement-settings'
			),
			{
				status: 200,
				body: fixtures.defaultEnhancedMeasurementSettings,
			}
		);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry.dispatch( CORE_FORMS ).setValues( FORM_SETUP, {
			autoSubmit: true,
			webDataStreamName: fixtures.createWebDataStream.displayName,
		} );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		fetchMock.post( REGEX_REST_GA4_CREATE_PROPERTY, {
			status: 200,
			body: properties[ 0 ],
		} );

		fetchMock.post( REGEX_REST_GA4_CREATE_WEBDATASTREAM, {
			status: 200,
			body: webDataStreams[ 2 ],
		} );

		fetchMock.get( REGEX_REST_GA4_ACCOUNT_SUMMARIES, {
			status: 200,
			body: accountSummaries,
		} );

		muteFetch( REGEX_REST_GA4_SETTINGS );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Complete setup/i } )
		).toBeInTheDocument();

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		expect( fetchMock ).toHaveFetchedTimes(
			1,
			REGEX_REST_GA4_CREATE_PROPERTY
		);
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			REGEX_REST_GA4_CREATE_WEBDATASTREAM
		);
		expect( fetchMock ).toHaveFetchedTimes( 1, REGEX_REST_GA4_SETTINGS );

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form only once in the case of an error', async () => {
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries( {
			accountSummaries: accountSummaries.accountSummaries.map(
				( account ) => ( {
					...account,
					propertySummaries: [],
				} )
			),
			nextPageToken: null,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry.dispatch( CORE_FORMS ).setValues( FORM_SETUP, {
			autoSubmit: true,
			webDataStreamName: fixtures.createWebDataStream.displayName,
		} );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		fetchMock.post( REGEX_REST_GA4_CREATE_PROPERTY, {
			status: 403,
			body: {
				code: 403,
				error: 'Insufficient permissions',
			},
		} );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Complete setup/i } )
		).toBeInTheDocument();

		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
		await waitForRegistry();
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Create property should have only been called once.
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			REGEX_REST_GA4_CREATE_PROPERTY
		);
		// Setup was not successful, so the finish function should not be called.
		expect( finishSetup ).not.toHaveBeenCalled();
		// Expect a console error due to the API error (otherwise this test will fail).
		expect( console ).toHaveErrored();
	} );

	describe.each( [
		[
			'account is changed',
			( getByRole ) =>
				fireEvent.click(
					getByRole( 'menuitem', {
						name: /Example Org/i,
						hidden: true,
					} )
				),
		],
		[
			'property is changed',
			( getByRole ) =>
				fireEvent.click(
					getByRole( 'menuitem', {
						name: /set up a new property/i,
						hidden: true,
					} )
				),
		],
		[
			'web data stream is changed',
			( getByRole ) =>
				fireEvent.click(
					getByRole( 'menuitem', {
						name: /set up a new web data stream/i,
						hidden: true,
					} )
				),
		],
	] )( 'when the %s', ( _, triggerChange ) => {
		beforeEach( async () => {
			muteFetch( { query: { tagverify: '1' } } );

			registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
			registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveGetAccountSummaries( accountSummaries );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveGetProperty( properties[ 0 ], {
					propertyID,
				} );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
					propertyIDs: [ propertyID ],
				} );
			registry
				.dispatch( MODULES_ANALYTICS_4 )
				.receiveGetEnhancedMeasurementSettings(
					{
						...defaultEnhancedMeasurementSettings,
						streamEnabled: false,
					},
					{
						propertyID,
						webDataStreamID,
					}
				);

			await registry
				.dispatch( MODULES_ANALYTICS_4 )
				.selectAccount( accountID );

			const { getByRole, getByLabelText, waitForRegistry } = render(
				<SetupForm />,
				{
					registry,
				}
			);
			await waitForRegistry();

			const switchControl = getByLabelText(
				'Enable enhanced measurement'
			);

			switchControl.click();

			const isEnhancedMeasurementEnabled = registry
				.select( CORE_FORMS )
				.getValue(
					ENHANCED_MEASUREMENT_FORM,
					ENHANCED_MEASUREMENT_ENABLED
				);

			expect( isEnhancedMeasurementEnabled ).toBe( false );

			act( () => {
				triggerChange( getByRole );
			} );

			await waitForRegistry();
		} );

		it( 'should revert the enhanced measurement from off to on', () => {
			const updatedIsEnhancedMeasurementEnabled = registry
				.select( CORE_FORMS )
				.getValue(
					ENHANCED_MEASUREMENT_FORM,
					ENHANCED_MEASUREMENT_ENABLED
				);

			expect( updatedIsEnhancedMeasurementEnabled ).toBe( true );
		} );
	} );
} );
