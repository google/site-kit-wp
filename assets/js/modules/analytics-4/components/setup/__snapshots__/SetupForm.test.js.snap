// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupForm renders the form correctly 1`] = `
<div>
  <form
    class="googlesitekit-analytics-setup__form"
  >
    <p
      class="googlesitekit-margin-bottom-0"
    >
      Please select the account information below. You can change this later in your settings.
    </p>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-select googlesitekit-analytics__select-account mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="987654321"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Example Com
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-0"
              style=""
            >
              Account
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select googlesitekit-analytics-4__select-property googlesitekit-analytics-4__select-property--loaded mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="1122334455"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Example Property (1122334455)
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-1"
              style=""
            >
              Property
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select googlesitekit-analytics-4__select-webdatastream mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="4000"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Main datastream (G-155BC2366E)
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-2"
              style=""
            >
              Web Data Stream
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
    </div>
    <div
      class="googlesitekit-analytics-enable-enhanced-measurement"
    >
      <div
        aria-checked="true"
        class="mdc-switch mdc-switch--checked"
        role="switch"
        tabindex="0"
      >
        <div
          class="mdc-switch__track"
        >
          Â 
        </div>
        <div
          class="mdc-switch__thumb-underlay"
        >
          <div
            class="mdc-switch__thumb"
          >
            <input
              checked=""
              class="mdc-switch__native-control"
              id="googlesitekit-switch-0"
              readonly=""
              role="switch"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <label
        class=""
        for="googlesitekit-switch-0"
      >
        Enable enhanced measurement
      </label>
      <p
        class="googlesitekit-module-settings-group__helper-text"
      >
        This allows you to measure interactions with your content (e.g. file downloads, form completions, video views). 
        <a
          aria-label="Learn more (opens in a new tab)"
          class="googlesitekit-cta-link"
          href="https://support.google.com/analytics/answer/9216061?hl=en-US"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Learn more
          </span>
          <span
            class="googlesitekit-icon-wrapper"
            style="margin-left: 5px;"
          >
            <svg />
          </span>
        </a>
      </p>
    </div>
    <div
      class="googlesitekit-setup-module__action"
    >
      <button
        class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
        disabled=""
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Complete setup
        </span>
      </button>
    </div>
  </form>
</div>
`;
